Fix flipping health_check_path on GA

aquarhead opened the pull request:
- Remove default value for `health_check_path`
- Update documentation
- Skip setting the resource data field for TCP GAs

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request


```release-note
Remove default value of `health_check_path`
```

Output from acceptance testing:

```
❯ make testacc TEST=./aws TESTARGS='-run=TestAccAwsGlobalAcceleratorEndpointGroup_tcp'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsGlobalAcceleratorEndpointGroup_tcp -timeout 120m
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_tcp
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_tcp (312.50s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	312.561s

❯ make testacc TEST=./aws TESTARGS='-run=TestAccAwsGlobalAcceleratorEndpointGroup'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsGlobalAcceleratorEndpointGroup -timeout 120m
go: downloading github.com/aws/aws-sdk-go v1.29.16
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_update
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_update
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_update
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_basic (245.99s)
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_tcp (281.40s)
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_update (307.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	307.279s
```

I'm proposing this change because for TCP endpoint groups there's no `health_check_path`:

![スクリーンショット 2020-03-05 11 48 32](https://user-images.githubusercontent.com/222272/75978897-62653180-5ed7-11ea-9f7f-701415b3c641.png)

With previous code, `terraform plan` will attempt to add a default `/` every time:

![image](https://user-images.githubusercontent.com/222272/75979078-aeb07180-5ed7-11ea-96e0-32c294ae0064.png)

But AWS doesn't care about it so it keeps try to add the field forever.

ewbankkit commented:
Relates: https://github.com/terraform-providers/terraform-provider-aws/issues/9329.
Relates: https://github.com/terraform-providers/terraform-provider-aws/issues/12882.

aquarhead commented:
Thanks for linking related issues. I wasn't able to find the previous one when I submit my PR. Hopefully there will be some traction now that multiple issues exists.

ewbankkit commented:
@aquarhead Yes, we are hoping to make progress on some of the non-[Core](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CORE_SERVICES.md) services.
It may be worth considering bflad's [suggestion](https://github.com/terraform-providers/terraform-provider-aws/issues/9329#issuecomment-571128685).  I think for HTTP(S) endpoints, removing `/` as the default `health_check_path` value will cause any configuration that were assuming that to now show a continual diff.

aquarhead commented:
So should `Default: "/"` be left untouched? And add "Computed: true"? Not sure how the resource update function should behave in that case... Wouldn't it make the statefile think the resource has a `health_check_path` of `/` while in reality it does not?

If `Computed: true` is to replace `Default: "/"` wouldn't that be the same situation as this one (in regards to causing diff) ? Or does it behave differently for existing resources? (Sorry I'm not familiar with these options)

ewbankkit commented:
@aquarhead Remove `Default:  "/",` and add `Computed: true`. Then if no value is specified by the user in their configuration, whatever value `DescribeEndpointGroup` returns for `health_check_path` will not show as a diff.

aquarhead commented:
Added, do I need to change other if statement?

ewbankkit commented:
@aquarhead Yes, there should be no need for those additional if statements now, but of course use the acceptance tests to verify. Thanks.

aquarhead commented:
Just removed all `if` and `GetOkExists` changes and also rebased to current master, now there's an error with acceptance test:

```
    TestAccAwsGlobalAcceleratorEndpointGroup_tcp: testing.go:684: Step 0 error: Check failed: Check 3/8 error: aws_globalaccelerator_endpoint_group.example: Attribute 'health_check_path' found when not expected
--- FAIL: TestAccAwsGlobalAcceleratorEndpointGroup_tcp (207.73s)
```

Seems `GetOk` will inject some value by default? Should I leave `GetOkExists` in place?

Actually all I care is that TF does not consider itself needs to update any values because of `health_check_path` is changed. Is there any particular technique I can use to specifically check that `health_check_path` does not trigger an update? (instead of checking values after `apply`)

ewbankkit commented:
@aquarhead After every acceptance test step effectively a `terraform plan` is run and by default any reported diffs will report as an error, so a successful test will verify that there is no drift.
For the currently failing test it is OK to have the `health_check_path` attribute set.
Yes `GetOk` will return the "zero" value, an empty string.

aquarhead commented:
@ewbankkit Thanks for that information, according to that I think just running a single TCP endpoint group creation test is enough, I've rebased all the changes and onto (this morning's) master. And acceptance test passes too:

```
<$> make testacc TEST=./aws TESTARGS='-run=TestAccAwsGlobalAcceleratorEndpointGroup_tcp'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsGlobalAcceleratorEndpointGroup_tcp -timeout 120m
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_tcp
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_tcp (244.77s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	244.834s
```

ewbankkit commented on website/docs/r/globalaccelerator_endpoint_group.html.markdown:
```diff
@@ -30,7 +30,7 @@ The following arguments are supported:
 * `endpoint_group_region` (Optional) - The name of the AWS Region where the endpoint group is located.
 * `health_check_interval_seconds` - (Optional) The time—10 seconds or 30 seconds—between each health check for an endpoint. The default value is 30.
-* `health_check_path` - (Optional) If the protocol is HTTP/S, then this specifies the path that is the destination for health check targets. The default value is slash (/).
+* `health_check_path` - (Optional) If the protocol is HTTP/S, then this specifies the path that is the destination for health check targets.
```
Let's add back `/` as the default for HTTP(S) (as it is 😄) and add a note about how changing to `Computed` affects drift detection.

```suggestion
* `health_check_path` - (Optional) If the protocol is HTTP/S, then this specifies the path that is the destination for health check targets. The default value is slash (`/`). Terraform will only perform drift detection of its value when present in a configuration.
```

aquarhead replied:
Hrm, but we don't have `Default: "/"` now? So it really depends on what AWS side decide to put as default. Or at least limit that to "The default value for HTTP/HTTPS ..."

Agree on the drift detection part.

ewbankkit replied:
I verified that AWS do return `/` if nothing is specified on creation (for HTTP(S) health check protocol).

ewbankkit replied:
You can decide on the wording 😄.

aquarhead replied:
@ewbankkit Given the sentence started with "If the protocol is HTTP/S" I felt yours read just fine :D

I was thinking the drift detection part might be reword like this "Terraform will only perform drift detection for this value if changed by the user.", but I'm not sure that's entirely correct (?)

ewbankkit replied:
I don't think we need to mention `TCP` as that's kind of implicit from not `HTTP/S`.
There seems to be no consistency in the existing documentation over the drift detection wording; I see

* `A configuration must be present for this argument to perform drift detection`
* `Terraform will only perform drift detection of its value when present in a configuration`
* `Terraform will only perform drift detection if a configuration value is provided`
* `Must be configured to perform drift detection`
* `To enable drift detection, must be configured`

So I suggest picking one of the above to avoid a 6th 😄.

aquarhead replied:
Totally agree.

Ehm.......... I thought I've incorporated your suggested changes.... let me check what happened to my `git` :facepalm:

ewbankkit commented:
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAwsGlobalAcceleratorEndpointGroup_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsGlobalAcceleratorEndpointGroup_ -timeout 120m
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_update
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_update
=== RUN   TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== PAUSE TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_basic
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_tcp
=== CONT  TestAccAwsGlobalAcceleratorEndpointGroup_update
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_basic (215.15s)
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_tcp (216.12s)
--- PASS: TestAccAwsGlobalAcceleratorEndpointGroup_update (240.11s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	240.215s
```

@aquarhead One minor tweak to the documentation and then this looks good. Thanks.

ewbankkit commented:
Closes #9329.

aquarhead commented:
Do I need to tweak the "Release note for CHANGELOG:" section in the [OP](https://github.com/terraform-providers/terraform-provider-aws/pull/12271#issue-384235136) ?

ewbankkit commented:
CHANGELOG looks OK.

aquarhead commented:
@ewbankkit Sorry, in the afternoon I thought I've added the doc change.... Just added now according to our discussion
